core/src/test/scala/unit/kafka/server/BrokerEpochIntegrationTest.scala

```diff
@@ -0,0 +1,298 @@
+      {
+        val partitionStates = Map(
+          tp -> new LeaderAndIsrRequest.PartitionState(controllerEpoch, brokerId2, LeaderAndIsr.initialLeaderEpoch + 1,
+            Seq(brokerId1, brokerId2).map(Integer.valueOf).asJava, LeaderAndIsr.initialZKVersion,
+            Seq(0, 1).map(Integer.valueOf).asJava, false)
+        )
+        val requestBuilder = new LeaderAndIsrRequest.Builder(
+          ApiKeys.LEADER_AND_ISR.latestVersion, controllerId, controllerEpoch,
+          broker2.kafkaController.brokerEpoch, // Correct broker epoch
+          partitionStates.asJava, nodes.toSet.asJava)
+
+        sendAndVerifySuccessfulResponse(controllerChannelManager, ApiKeys.LEADER_AND_ISR, requestBuilder)
+        TestUtils.waitUntilLeaderIsKnown(Seq(broker2), tp, 10000)
+      }
+
+      // Send UpdateMetadata request with correct broker epoch
+      {
+        val partitionStates = Map(
+          tp -> new UpdateMetadataRequest.PartitionState(controllerEpoch, brokerId2, LeaderAndIsr.initialLeaderEpoch + 1,
+            Seq(brokerId1, brokerId2).map(Integer.valueOf).asJava, LeaderAndIsr.initialZKVersion,
+            Seq(0, 1).map(Integer.valueOf).asJava, Seq.empty.asJava)
+        )
+        val liverBrokers = brokers.map { broker =>
+          val securityProtocol = SecurityProtocol.PLAINTEXT
+          val listenerName = ListenerName.forSecurityProtocol(securityProtocol)
+          val node = broker.node(listenerName)
+          val endPoints = Seq(new EndPoint(node.host, node.port, securityProtocol, listenerName))
+          new UpdateMetadataRequest.Broker(broker.id, endPoints.asJava, broker.rack.orNull)
+        }
+        val requestBuilder = new UpdateMetadataRequest.Builder(
+          ApiKeys.UPDATE_METADATA.latestVersion, controllerId, controllerEpoch,
+          broker2.kafkaController.brokerEpoch, // Correct broker epoch
+          partitionStates.asJava, liverBrokers.toSet.asJava)
+
+        sendAndVerifySuccessfulResponse(controllerChannelManager, ApiKeys.UPDATE_METADATA, requestBuilder)
+        TestUtils.waitUntilMetadataIsPropagated(Seq(broker2), tp.topic(), tp.partition(), 10000)
+        assertEquals(brokerId2,
+          broker2.metadataCache.getPartitionInfo(tp.topic(), tp.partition()).get.basePartitionState.leader)
+      }
+
+      // Send StopReplica request with stale broker epoch
+      {
+        val requestBuilder = new StopReplicaRequest.Builder(
+          ApiKeys.STOP_REPLICA.latestVersion, controllerId, controllerEpoch,
+          broker2.kafkaController.brokerEpoch, // Correct broker epoch
+          true, Set(tp).asJava)
+
+        sendAndVerifySuccessfulResponse(controllerChannelManager, ApiKeys.STOP_REPLICA, requestBuilder)
+        assertTrue(broker2.replicaManager.getPartition(tp).isEmpty)
+      }
+    } finally {
+      controllerChannelManager.shutdown()
+      metrics.close()
+    }
+  }
+
+  @Test
+  def testControlRequestWithStaleBrokerEpoch() {
```

Comment:
The code in this method is quite similar to that in testControlRequestWithCorrectBrokerEpoch(). Should we merged them somehow?

Reply:
Sure. I have merged them into a single function.